src/NServiceBus.Metrics/RawData/RawDataReporter.cs

```diff
@@ -0,0 +1,126 @@
+        readonly BinaryWriter writer;
+        readonly MemoryStream memoryStream;
+        readonly CancellationTokenSource cancellationTokenSource;
+        readonly TimeSpan maxSpinningTime;
+        Task reporter;
+
+        static readonly TimeSpan DefaultMaxSpinningTime = TimeSpan.FromSeconds(5);
+        static readonly TimeSpan singleSpinningTime = TimeSpan.FromMilliseconds(50);
+        static ILog log = LogManager.GetLogger<RawDataReporter>();
+
+        public RawDataReporter(IDispatchMessages dispatcher, string destination, HostInformation hostInformation, RingBuffer buffer, string messageTypeName, string endpointName,
+            WriteOutput outputWriter) : this(dispatcher, destination, hostInformation, buffer, messageTypeName, endpointName, outputWriter, DefaultFlushSize, DefaultMaxSpinningTime)
+        { }
+
+        public RawDataReporter(IDispatchMessages dispatcher, string destination, HostInformation hostInformation, RingBuffer buffer, string messageTypeName, string endpointName,
+            WriteOutput outputWriter, int flushSize, TimeSpan maxSpinningTime)
+        {
+            this.buffer = buffer;
+            this.flushSize = flushSize;
+            this.maxSpinningTime = maxSpinningTime;
+            this.outputWriter = entries => outputWriter(entries, writer);
+            this.dispatcher = dispatcher;
+            this.destination = new UnicastAddressTag(destination);
+
+            headers[Headers.OriginatingMachine] = RuntimeEnvironment.MachineName;
+            headers[Headers.OriginatingHostId] = hostInformation.HostId.ToString("N");
+            headers[Headers.EnclosedMessageTypes] = "NServiceBus.Metrics." + messageTypeName;
+            headers[Headers.OriginatingEndpoint] = endpointName;
+
+            memoryStream = new MemoryStream();
+            writer = new BinaryWriter(memoryStream);
+            cancellationTokenSource = new CancellationTokenSource();
+        }
+
+        public void Start()
+        {
+            reporter = Task.Run(async () =>
+            {
+                while (cancellationTokenSource.IsCancellationRequested == false)
+                {
+                    var totalSpinningTime = TimeSpan.Zero;
+
+                    // spin till either MaxSpinningTime is reached OR items to consume are more than FlushSize
+                    while (totalSpinningTime < maxSpinningTime)
+                    {
+                        if (cancellationTokenSource.IsCancellationRequested)
+                        {
+                            break;
+                        }
+
+                        var itemsToConsume = buffer.RoughlyEstimateItemsToConsume();
+                        if (itemsToConsume >= flushSize)
+                        {
+                            break;
+                        }
+
+                        totalSpinningTime += singleSpinningTime;
+                        await Task.Delay(singleSpinningTime).ConfigureAwait(false);
```

Comment:
Funny. @danielmarbach mentioned that it could be picked up ;-)
Yes. The loop deals with it checking token in an `if` possibly delaying stop by `singleSpinningTime`. Is it bad?